src/sequence/mod.rs

```diff
@@ -719,22 +723,162 @@ mod tests {
     }
 
     // verify data convergence on a set of replicas and with the expected length
-    fn verify_data_convergence(replicas: &[&Sequence], expected_len: u64) {
+    fn verify_data_convergence(replicas: Vec<Sequence>, expected_len: u64) {
         // verify replicas have the expected length
         // also verify replicas failed to get with index beyond reported length
         let index_beyond = SequenceIndex::FromStart(expected_len);
-        for r in replicas {
+        for r in &replicas {
             assert_eq!(r.len(), expected_len);
             assert_eq!(r.get(index_beyond), None);
         }
 
         // now verify that the items are the same in all replicas
         for i in 0..expected_len {
             let index = SequenceIndex::FromStart(i);
-            let r0_entry = replicas[0].get(index);
-            for r in replicas {
-                assert_eq!(r0_entry, r.get(index));
+            let r0_entry = &replicas[0].get(index);
+            for r in &replicas {
+                assert_eq!(r0_entry, &r.get(index));
             }
         }
     }
+
+    /// Generate a vec of Sequence replicas of some length
+    fn generate_replicas(quantity: i32, xorname: XorName, tag: u64) -> Result<Vec<Sequence>> {
+        let mut replicas: Vec<Sequence> = vec![];
+        let owner = gen_public_key();
+
+        for _ in 0..quantity {
+            let mut seq = Sequence::new_public(owner, xorname, tag);
+            let perms = BTreeMap::default();
+            let owner_op = seq.set_public_policy(owner, perms)?;
+            seq.apply_public_policy_op(owner_op)?;
+            replicas.push(seq)
+        }
+
+        Ok(replicas)
+    }
+
+    proptest! {
+        #[test]
+        fn proptest_seq_doesnt_crash_with_random_data(s in "\\PC*".prop_map(|s| s.into_bytes())) {
+            let actor1 = gen_public_key();
+            let sequence_name = XorName::random();
+
+            let sdata_tag = 43_001u64;
+
+            // Instantiate the same Sequence on two replicas with two diff actors
+            let mut replica1 = Sequence::new_public(actor1, sequence_name, sdata_tag);
+            let mut replica2 = Sequence::new_public(actor1, sequence_name, sdata_tag);
+
+            // Set Actor1 as the owner
+            let perms = BTreeMap::default();
+            let owner_op = replica1.set_public_policy(actor1, perms)?;
+            replica2.apply_public_policy_op(owner_op)?;
+
+            // Append an item on replicas
+            let append_op1 = replica1.append(s.clone())?;
+            let append_op2 = replica2.append(s)?;
+
+            replica1.apply_data_op(append_op1)?;
+            replica2.apply_data_op(append_op2)?;
+
+            verify_data_convergence(vec![replica1, replica2], 1);
```

Comment:
I think this test should fail? we have appended the same bytes to both replicas, but they are different items, perhaps we can use different bytes for appending to each replica so we can easily differentiate them?.
Also, we haven't apply the operation from replica1 to replica2, and vice versa, which should be the normal flow and then the convergence check should expect two items on each replica.

Reply:
Aha, I was thinking our `verify_convergence` was doing more than it is.